### PR TITLE
Gearboxing memcopies

### PR DESCRIPTION
**Pull request by definelicht**
Allow copying between container of different memory widths on the FPGA (#228).

### CODE REVIEW

File: dace/codegen/targets/fpga.py [hunk 1]
```diff
@@ -101,6 +101,11 @@ def __init__(self, frame_codegen, sdfg):
             dace.dtypes.StorageType.CPU_Stack,
             dace.dtypes.StorageType.FPGA_Global, None, self)
 
+        # Inspect the vector length of all memlets leading to each memory, to
+        # make sure that they're consistent, and to allow us to instantiate the
+        # memories as vector types to enable HLS to generate wider data paths.
+        self._memory_widths = type(self).detect_memory_widths(sdfg)
```
Comment (tbennun) [inline dace/codegen/targets/fpga.py:107]: isn't just "self" enough?

File: dace/codegen/targets/intel_fpga.py [hunk 2]
```diff
@@ -987,6 +1010,32 @@ def generate_undefines(self, sdfg, dfg, node, callsite_stream):
                     callsite_stream.write("#undef {}".format(memlet_name),
                                           sdfg, sdfg.node_id(dfg), node)
 
+    def generate_converter(self, is_unpack, dtype, veclen, node, state_id,
+                           sdfg, function_stream):
+        if (is_unpack, dtype, veclen) in self.converters_generated:
+            return
+        if is_unpack:
+            function_stream.write(
+                """\
```
Comment (tbennun) [inline dace/codegen/targets/intel_fpga.py:1019]:
```suggestion
                """
```

File: dace/codegen/targets/intel_fpga.py [hunk 3]
```diff
@@ -987,6 +1010,32 @@ def generate_undefines(self, sdfg, dfg, node, callsite_stream):
...
+void unpack_{dtype}{veclen}(const {dtype}{veclen} value, {dtype} *const ptr) {{
+    #pragma unroll
+    for (int u = 0; u < {veclen}; ++u) {{
+        ptr[u] = value[u];
+    }}
+}}\n\n""".format(dtype=dtype, veclen=veclen), sdfg, state_id, node)
+        else:
+            function_stream.write(
+                """\
```
Comment (tbennun) [inline dace/codegen/targets/intel_fpga.py:1028]:
```suggestion
                """
```

File: dace/runtime/include/dace/xilinx/access.h [hunk 4]
```diff
@@ -71,4 +71,24 @@ void Write(StreamView<T, vector_length, capacity> &stream_view,
   stream_view.push(value);
 }
 
+template <typename T, unsigned vector_length>
```
Comment (tbennun) [inline dace/runtime/include/dace/xilinx/access.h:74]: isn't this redundant with the opencl-generated pack/unpack?

File: dace/codegen/targets/fpga.py [hunk 5]
```diff
@@ -677,6 +685,41 @@ def _emit_copy(self, sdfg, state_id, src_node, src_storage, dst_node,
 
             ctype = src_node.desc(sdfg).dtype.ctype
 
+            # For a single vector access, tolerate not having set a range
+            if copy_shape[-1] != 1:
+                # Adjust for vectorization length
+                copy_shape[-1] = copy_shape[-1] / memlet.veclen
```
Comment (tbennun) [inline dace/codegen/targets/fpga.py:691]: It's kinda weird that (a) we have a special case for size 1 and (b) we need to divide by memlet veclen but nothing to do I guess...

Also, maybe integer division? //

File: dace/codegen/targets/fpga.py (hunk 1 above)
Comment (definelicht) [reply dace/codegen/targets/fpga.py:107]: yes :-)

File: dace/codegen/targets/intel_fpga.py (hunk 2 above)
Comment (definelicht) [reply dace/codegen/targets/intel_fpga.py:1019]: Nah, the convention is that you clear space after yourself, not before yourself.

File: dace/codegen/targets/intel_fpga.py (hunk 3 above)
Comment (definelicht) [reply dace/codegen/targets/intel_fpga.py:1028]: Nah, the convention is that you clear space after yourself, not before yourself.

File: dace/runtime/include/dace/xilinx/access.h (hunk 4 above)
Comment (definelicht) [reply dace/runtime/include/dace/xilinx/access.h:74]: This is Xilinx, vectors are defined completely differently, so we don't use the same functions

File: dace/codegen/targets/fpga.py (hunk 1 above)
Comment (definelicht) [reply dace/codegen/targets/fpga.py]: fixed

**Pull request merged**